refactor: use RelayTransaction in BroadcastTransaction utility

ariard opened the pull request:
Implementing suggestion in https://github.com/bitcoin/bitcoin/pull/15713#discussion_r306571420.

Seems a reason of these node utilities is to glue with already there functions, so we should reuse them.

MarcoFalke commented:
re-run ci

jnewbery commented:
Concept ACK.

I had some suggested changes to `BroadcastTransaction()` here: https://github.com/jnewbery/bitcoin/commit/ecfeea015919cc44b465e969e65cb2c459c1235c (removing unnecessary local variables, improving comments). Feel free to take that commit if you think it makes sense to bundle it in this PR.

MarcoFalke commented on src/net_processing.h:
```diff
@@ -90,4 +90,7 @@ struct CNodeStateStats {
 bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats);
 
+/** Relay transaction to every node */
+void RelayTransaction(const CTransaction& tx, CConnman* connman);
```
style-nit: connman can be a const reference instead of a mutable raw pointer, since we don't modify connman, but push the tx to the set of all nodes.

```suggestion
void RelayTransaction(const CTransaction& tx, const CConnman& connman);
```

MarcoFalke replied:
style-nit: (feel free to ignore)

It is sufficient to pass the txid (as we assume the tx is in the mempool). So maybe make the signature similar to `ChainImpl`'s `void relayTransaction(const uint256& txid)` and use it there as well?

maflcko reviewed (approved):
ACK ddab0452020bba2dcdd342437454cd01227d401e

promag commented:
ACK, also Marco suggestions.

ariard commented:
@MarcoFalke addressed your comments but john commit makes more sense than I take it on top of #15713

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1305,7 +1305,7 @@ bool static AlreadyHave(const CInv& inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
-static void RelayTransaction(const CTransaction& tx, CConnman* connman)
+void RelayTransaction(const uint256& txid, CConnman* connman)
 {
     CInv inv(MSG_TX, tx.GetHash());
```
```suggestion
    CInv inv(MSG_TX, txid);
```

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -2480,7 +2480,9 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
 
-        CInv inv(MSG_TX, tx.GetHash());
+
+        const uint256& hash = tx.GetHash();
```
`GetHash` already returns a reference, so you don't need this alias, I think

MarcoFalke commented on src/net_processing.h:
```diff
@@ -90,4 +90,7 @@ struct CNodeStateStats {
 bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats);
 
+/** Relay transaction to every node */
+void RelayTransaction(const CTransaction& tx, const CConnman& connman);
```
```suggestion
void RelayTransaction(const uint256& txid, const CConnman& connman);
```

MarcoFalke commented:
re-ACK apart from compile failures

Commit:
refactor : use RelayTransaction in BroadcastTransaction utility

To do so, we also refactor RelayTransaction to take a txid
instead of passing a tx

ariard commented:
Sorry for compilation failures, should be good now.

MarcoFalke commented on src/node/transaction.cpp:
```diff
@@ -69,10 +70,7 @@ TransactionError BroadcastTransaction(const CTransactionRef tx, uint256& hashTx,
-    g_connman->ForEachNode([&inv](CNode* pnode) {
-        pnode->PushInventory(inv);
-    });
+    RelayTransaction(hashTx, std::move(*g_connman));
```
I don't think you can move references

```suggestion
    RelayTransaction(hashTx, *g_connman);
```

Also, what happened to this suggestion:

```diff
diff --git a/src/interfaces/chain.cpp b/src/interfaces/chain.cpp
index 02f39cef8e..22e4aaedaf 100644
--- a/src/interfaces/chain.cpp
+++ b/src/interfaces/chain.cpp
@@ -9,6 +9,7 @@
 #include <interfaces/handler.h>
 #include <interfaces/wallet.h>
 #include <net.h>
+#include <net_processing.h>
 #include <node/coin.h>
 #include <policy/fees.h>
 #include <policy/policy.h>
@@ -292,8 +293,7 @@ public:
     }
     void relayTransaction(const uint256& txid) override
     {
-        CInv inv(MSG_TX, txid);
-        g_connman->ForEachNode([&inv](CNode* node) { node->PushInventory(inv); });
+        RelayTransaction(txid, *g_connman);
     }
     void getTransactionAncestry(const uint256& txid, size_t& ancestors, size_t& descendants) override
     {

ariard replied:
You can move reference but that's useless, drop it so.

Added suggestion, didn't see it at first

promag commented on src/net_processing.cpp:
```diff
@@ -1305,10 +1305,10 @@ bool static AlreadyHave(const CInv& inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
-    CInv inv(MSG_TX, tx.GetHash());
-    connman->ForEachNode([&inv](CNode* pnode)
+    CInv inv(MSG_TX, txid);
+    connman.ForEachNode([&inv](CNode* pnode)
```
µnit, pardon
```cpp
[&inv](CNode* pnode) {
```

MarcoFalke replied:
I'd rather keep it the way it is now

promag reviewed (commented):
ACK 9bc8b28c1d26c28edf4bbc890be97c0ad7a73cb9, verified there are no more `PushInventory(CInv(MSG_TX, ...`, nice refactor, :+1: @amitiuttarwar.

MarcoFalke commented:
ACK 9bc8b28c1d26c28edf4bbc890be97c0ad7a73cb9

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 9bc8b28c1d26c28edf4bbc890be97c0ad7a73cb9
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjNYgv/ecqESlAxc75rOwly8VCGKZjH4Zdb0+lfb3LnFTNgsPkd2OT1UZ9t3qby
J18SGBradFqTrRIivGqwxnOkIKqc2HO+7g2M2NERzdudVE0TOauFIMC+R0avTpcC
xRetRqa7jA445c/r0JwaOXSaS84KTXEY9A7X1Me29WyAPZoxStnJajp7z4ZAiezy
WUtk5J0JTbRu9Hf1DqRKFKMmywnWMREmJ/zuOHHoYDgLQ2yLzfKgcAn3z853hcno
uWJ2tlY53i7L2YrM9oi8XcxjQ4RpLPB57uHGvc3RKNGFfmtB20+8OFKsbvCwpyQM
cbl1NeWW/Cog8e8aoSlJGLsdfET90E6xYfGJIlxsRaN3QG7BLRK+Y72+jOgLulkQ
j3d8/rwloWRW5Iz/nDVMM5zj6DmbwWCiYxWGZk9ygNvY4Jh+8i9d5Wa78Z6p1BHU
V2RaKA/reYduiEd/ds4oPjpSq3JFCFPopVdOxQGbCx0VlDZKodT1s1yffVE8SqAE
r37wmoMd
=mYsj
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `55d66b1be7c55577a29116b202deff842019db2cbb0adf69487cb6e652257741  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010855d66b1be7c55577a29116b202deff842019db2cbb0adf69487cb6e652257741f0102fa213bd130d9205af45ea278a92b9f908fff0104908e280256eefc4f6f20db8598896b008f1045d399888f008cfa3b02e667f2af60083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0105022ed0a1f2b58510de3ef70b2e0f3c808f0207a12c0a9da060da27afcb0f82af9916f43c64f51911cfb02bcd26526bd5e48a308f1045d399888f0081e4d0f89dbef6d210083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105fa7483ef04abccfa3b8e02c020f66b008f02041994988cf284e0c150079e37b34fc1dccbb205ed9cd7a0a33b46e354ddfe30c08f1045d399889f0087e6855f6a1397bea0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010d464337ec4b254f714bc2f706f7cfff708f020fd156f7a4d7fe9ca375490132a2db3badd27388ed2b96d77f6965a0eb2c89b7d08f1045d399888f008c274ff39345a3b210083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

jnewbery commented on src/net_processing.h:
```diff
@@ -90,4 +90,7 @@ struct CNodeStateStats {
 bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats);
 
+/** Relay transaction to every node */
+void RelayTransaction(const uint256&, const CConnman& connman);
```
nit: Please add the name of the txid argument:

`void RelayTransaction(const uint256& txid, const CConnman& connman);`

jnewbery commented on src/node/transaction.cpp:
```diff
@@ -5,6 +5,7 @@
 
 #include <consensus/validation.h>
 #include <net.h>
```
nit: you can remove the `net.h` include now that `CNode` is not used in this file.

jnewbery commented on src/interfaces/chain.cpp:
```diff
@@ -9,6 +9,7 @@
 #include <interfaces/handler.h>
 #include <interfaces/wallet.h>
 #include <net.h>
```
nit: you can remove the `net.h` include now that this file doesn't use `CNode`

jnewbery reviewed (commented):
ACK 9bc8b28c1d26c28edf4bbc890be97c0ad7a73cb9

Three nits inline. Take 'em or leave 'em.

ariard commented:
thanks @jnewbery will correct the nits in one of my other PRs to not invalidate reviews

jonatack reviewed (commented):
ACK 9bc8b28c1d26c28edf4bbc890be97c0ad7a73cb9, second @jnewbery's suggestions, my guess is they could be added without risking delaying this PR.

jnewbery commented:
I think this is ready for merge. It could go in with or without https://github.com/jnewbery/bitcoin/commit/72f29c34c069d7cd87424f3c4dbefd4aa25400e0 (I'll open a quick follow-up PR if we merge this as is)

ariard commented:
Okay, will rebase #15713 soon after and address nit on it at same time

jonatack commented:
NP. FWIW I also built and tested with @jnewbery's changes and all pass.
```cpp
diff --git a/src/interfaces/chain.cpp b/src/interfaces/chain.cpp
index 22e4aaedaf..30966c50ad 100644
--- a/src/interfaces/chain.cpp
+++ b/src/interfaces/chain.cpp
@@ -8,7 +8,6 @@
 #include <chainparams.h>
 #include <interfaces/handler.h>
 #include <interfaces/wallet.h>
-#include <net.h>
 #include <net_processing.h>
 #include <node/coin.h>
 #include <policy/fees.h>
diff --git a/src/net_processing.h b/src/net_processing.h
index 1d26164b18..c60ed4b21f 100644
--- a/src/net_processing.h
+++ b/src/net_processing.h
@@ -91,6 +91,6 @@ struct CNodeStateStats {
 bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats);
 
 /** Relay transaction to every node */
-void RelayTransaction(const uint256&, const CConnman& connman);
+void RelayTransaction(const uint256& txid, const CConnman& connman);
 
 #endif // BITCOIN_NET_PROCESSING_H
diff --git a/src/node/transaction.cpp b/src/node/transaction.cpp
index 5cd567a5c4..ae3edd5d6a 100644
--- a/src/node/transaction.cpp
+++ b/src/node/transaction.cpp
@@ -4,7 +4,6 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <consensus/validation.h>
-#include <net.h>
 #include <net_processing.h>
 #include <txmempool.h>
 #include <util/validation.h>
```

MarcoFalke commented:
`net.h` might still be included through `net_processing.h`, so the follow-up changes are only stylistic
